### ISSUE TITLE
What about a PrivacyComponent/PrivacyBundle

### ISSUE THREAD

**Issue by Aerendir** [symfony/symfony#26231]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no

Hello to everyone,
I'm opening this issue as on 25 May the new [General Data Protection Regulation will be executive in all EU member states](http://eur-lex.europa.eu/legal-content/EN/TXT/?uri=uriserv:OJ.L_.2016.119.01.0001.01.ENG&toc=OJ:L:2016:119:TOC).

It is a great revolution in privacy and it is inspired by concepts and ideas as ["Privacy by default" and "Privacy by design"](https://en.wikipedia.org/wiki/Privacy_by_design).

Those are concepts that are deeply linked with the development cycle of an application and affect ALL applications we devlop.

Currently I'm not aware of a unified and simple way of handling the privacy in our PHP/Symfony apps.

**This issue is to open the discussion about the privacy in our applications and understand if may be interest in developing a component (and eventually a bundle) that tries to abstract the common activities required to make our applications privacy compliant.**

Privacy has relevant impact on our application, from the simple collection of emails to the even simpler showing of a Facebook Like button on our pages.

Here you can find an infographic from the EU that explains the GDPR: http://ec.europa.eu/justice/smedataprotect/index_en.htm

Here you can find the complete text of the GDPR: http://eur-lex.europa.eu/legal-content/EN/TXT/PDF/?uri=CELEX:32016R0679&from=IT

**I think at the component as a simple and abstract way to manage all those aspects.**

I imagine **something like `{% if privacy_accepted('facebook_like') %}` to use in our `Twig` templates** to display or not a like button. Or to **automatically add a checkbox to our forms** that has be checked before we can collect, process and save personal data.

But the discussion doesn't stop here.

**There are also all the issues related to cookies** that the European Union also regulates with an _ad hoc_ regulation ([the EU Cookie Law](http://ec.europa.eu/ipg/basics/legal/cookies/index_en.htm)).

This regulation states, besides the others, that on **each website must exist a page where all cookies installed are listed** and that the user has the right to refuse some or all of them in a very granular way (I saw an application of this on the ico.org.uk website but currently it seems they discontinued this possibility).

You can see sample page about cookies on the https://ico.org.uk/global/cookies/ page:

<img width="1256" alt="schermata 2018-02-19 alle 13 23 57" src="https://user-images.githubusercontent.com/1968622/36377666-36d46798-1578-11e8-9f6d-3a2b263b06af.png">

To manage the cookies they currently use a third party service but I think this approach is not 1. standard and 2. easy to use.

I think the entire process can be really simplified providing a component.

**HOW I IMAGINE THE FIRST DRAFT OF THE PRIVACY COMPONENT**

I imagine the Privacy Component work in a really similar way as the Workflow Component works:

`Officer` = `Workflow`
`PolicyBuilder` = `DefinitionBuilder`
`Policy` = `Definition`
`Consents` = `Marking` (contains all consents and withholds - denials: I don't know which is the more right term to use - of a visitor/user in `json`)

To use the component I imagine something like this:

```php
PolicyBuilder = new PolicyBuilder()
Policy = PolicyBuilder->addConsentRequest(‘google_analytics’)
->addConsentRequest(‘facebook’)
->drawUp();

Officer = new Officer(Policy, Consents)
Officer->consentedTo(‘google_analytics’, $user)
Officer->getPolicy():Policy
Officer->getConsents():Consents
Officer->getWithholds(): [to define]
Officer->setConsentTo(‘google_analytics)
```

Starting from here I'd like, now, to know what do you think about this, try to collect common use cases, try to understand current solutions you use to manage privacy in your applications, and, in general, anything that can be useful to make the handling of privacy a simpler thing t which we have to not to think to when developing Symfony applications...

Thank you for reading up to here :)

**UPDATE**

## The privacy and the PHP world

There are some interesting projects already in place that try to make developer's lives easier when dealing with privacy.

## [PHP-PrivacyPolicy](https://github.com/delight-im/PHP-PrivacyPolicy)

> Programmatically composable privacy policies for humans and machines

With a fluent interface it permits to write a privacy policy in various formats.

Also if the code can be really improved (currently is not so flexible to generate policies in different languages) and also if it, in my opinion, does too much, it bring an interesting idea: the possibility to write the policy in `Json`. This means that if we, for example, move our data to third party services, we may read their policy in Json and make our apps aware of the privacy consequences of this processing.

## [GDPR for WordPress](https://www.gdprwp.com/)

> We aim to give plugin developers a simple solution to GDPR validate their plugin, and offer Website Administrators the overview and tools to handle the administrative tasks involved with being GDPR compliant.

This is really interesting: it tries abstract privacy management in plugins to give a clear picture of what happens with personal data when using a particular plugin.

Is really close to what I'd like to see in Symfony.

For example, currently, I'm using FOSUserBundle in my application but I have no idea of which cookies it install, for how much time they are active on clients, why they are installed, which is their purpose...

Binding the cookie management with the `Session` it is possible to automatically generate a list of installed cookies: this list is MANDATORY and is asked for by lawyers who have to write a cookie policy where they list the cookies installed by an application.

Currently this means that we have to dig in the code and find all cookies installed.

If we activate a new bundle we have to repeat the process.

But, what if the bundle is updated and new cookies are installed or old cookies are not installed anymore?

I think a tool that can automatically generate a list of cookies that we, as developers, can pass to lawyers so they can include it in the cookies policies.

Something that is already done, for example, for translations using the JMSTranslationBundle: it reads the code finding all the calls to `Translator->trans()` and creates a list of missed translations.

**Consider that this is a major requirement also for projects depending on Symfony, like PrestaShop or Magento: they are particularly sensitive to the GDPR as merchant MUST fully comply!**

## OTHER INFORMATION REQUIRED BY LAWYERS

Lawyers want to also know where the data are stored, in which countries, which are the security measures put in place, ecc.

A lot of those information can be provided in an automatic way as they are striclty related to applications: for example, if we use AWS, then we have the AWS SDK installed through Composer: this may be intercepted.

They also want to know a lot of information about the time of the processing: for how much time does a cookie remain installed on the client? This is something set in the code directly and there MUST be a correspondence between what is wrote in the policies and what the application really does.

Those are only examples. I'm sure that growing the component (and the related bundle) there will be more and more opportunities for improvements.

But, as pointed out by some users (@stof and @sstok ) the best approach is to first develop the component outside Symfony and then include it in the framework.

So this is exactly what I'm going to do:

1. Write a condensed version of the GDPR that contain only the articles related to Privacy by Design and that a developer MUST know
2. Write a first draft of the component with some basic features (like the twig extension for example) from where to start to build more complex features.

**Comment by Xymanek** [comment]
If example code was an actual code it would be so much easier to understand it...

**Comment by stof** [comment]
I think this should first be developed outside Symfony to see whether it can work in a useful way, similar to what was done for Guard.
AFAICT, there is nothing forbidding to implement this outside the core first (as this is about a new component, not about something requiring to alter an existing component)

**Comment by Aerendir** [comment]
@Xymanek , first of all thank you for your reply! :)

The code is "pseudo" just to make clear what each method returns and give a general idea of how to use the component I'm imagining.

You don't fully understand it because, reading it again, there are some pseudo-bugs: for example, the `Officer` accepts a `Consents` that is not clear from where it comes.

But, basically the idea is this:

1. The `PolicyBuilder` class creates a builder to which is possible to pass simple strings hold by an internal array: nothing complex.
2. Calling the method  `PolicyBuilder::drawUp()` we get a `Policy` object that contains all the possible consents a User can give on our application;
3. The `Policy` is passed to the `Officer` when initializing it;
4. Each `User` has a field saved to the database in `json` format where are listed all the consents given or withhold
5. Calling `Officer::consentedTo('google_analytics', $user)` the `Officer` reads the `Consents` object created from the `json` consents associated to the `User` object and decides if the `User` gave or not his/her consent. (exactly has the `Workflow` component does to decide if a transition can be applied or not).

All the other method are just to access the internals.

As I told, this is only a draft, a superficial explanation of what I currently imagine.

But there is more and more room for improvements and discussion.

For this reason I'd like to know currently how we manage privacy in our application: while for, for example, for `Workflow` and state machines there are a lot of documents, papers and in general a lot of documentation from which take inspiration, for the privacy all those information are completely absent as each developer solves the problem on his/her own, considering it only a nuisance, without sharing any knowledge about it.

But now we cannot adopt this approach anymore as the privacy is raising as a pillar of each application, something to think about from start.

So, how do you manage privacy in you application currently? Are there any specific issue you commonly face? How do you currently solve them?

[ @javiereguiluz : what does the `RFC` tag mean?]

**Comment by Aerendir** [comment]
@stof , yes, maybe: this is exactly what I'd like to understand: which is the common path to follow? I first write the code in a repository of mine and then pass the code to Symfony? Or is the prcedure different?

Anyway, how do you currently manage the privacy related problems? What do you often face and how do you solve the privacy management in your applications?

**Comment by sstok** [comment]
RFC means Request for Commend. Which means that anyone is invited to respond and give suggestions (rather then this being a concrete suggestion). 👍

**Comment by Aerendir** [comment]
@sstok , yes, after posting I had a flash of inspiration! :P

Ok on developing this outside Symfony first.

> Hehe 😄 I fear most of us don't do this very strictly atm which is also why this law is introduced.
I hope to be wrong honestly.

Unfortunately you are right: I've posted a similar question to some groups related to PHP and the general reaction is of non-consciousness of the topic. More, there is a lot of confusion.

The problem, I think, is that read "legalese" is something really boring and often also complex: as if you pretend a lawyer being able to read a PHP code, but with just a bit of more "elasticity as the legal codes" are interpreted by humans and so they are prone to differences in concrete conclusions (while a program can only work or don't work).

Anyway this regulation is going to be really important and it will have deep impact on how we develop applications.

> but for the best insight we need someone who fully understands this new law and is able to concretely specify what is needed in technical terms 🧐

This is exactly the reason that pushed me to open this issue: I have a strong academic legal education and a strong private education on PHP.

I understand well both development and legal issues.

Privacy is one of my fields of interest and for this reason I'm putting the accent on this topic.

Since the Cookie Law I pointed out the urgency about an integrated way of handling cookies in application. Frameworks in every language should provide a way to manage them.

With this new regulation (that is immediately effective in each EU member state, without the requiring of an ad hoc law by each state as happens for directives) the field of application is really really broad: the regulation states a principle of application that goes outside the territoriality: namely, also US, Chinese, Indian,  and any other country in the world MUST respect the regulation if personal data of european people are involved.

In the mean time I've done some more researches and I've found a lot of interesting things: I'm going to update my initial post with the new resources I've found.

I'm also going to prepare a GDPR regulation for developers: in my intentions is a condensed version of the GDPR regulation that will contain only articles and subsections that MUST be read by a developer to better understand it.

Give me some time and I will post it somewhere on the web: it will be a good starting point to better understand the changes...

**Comment by Xymanek** [comment]
> For example, currently, I'm using FOSUserBundle in my application but I have no idea of which cookies it install, for how much time they are active on clients, why they are installed, which is their purpose

This point actually gives this idea a big incentive to be an sf component - bundles can contribute their information to a common standard

That said making a new sf component will likely take quite a while since everyone will have to agree on details...

> tool that can automatically generate a list of cookies 

Why limit to cookies? Things like FOSUserBundle can also contribute information about what they store in database

**Comment by Aerendir** [comment]
@Xymanek , yes, this is exactly why I'd like to integrate it in Symfony: to make every bundle creator able to define its privacy data processings clear and defined in a standard way.

I speak a lot about cookies as their regulation is older and so there are stronger opinions about them.

This new regulation really new and so not all the various consequences of its application are clear.

Anyway already there are some points that for sure can be covered by a tool like the one I'm proposing.

I'm going to read again the regulation with a focus on development to find more and more point of contacts and try to understand how a tool like this can help us as developers.

**Comment by Aerendir** [comment]
To ease the understanding of the GDPR I'm going to write it in a condensed form with only the relvant parts related to development.

I've also setup the repository that will contain the code of the component.

You can check it here: https://github.com/Aerendir/PHPrivacy

**Comment by stof** [comment]
> yes, this is exactly why I'd like to integrate it in Symfony: to make every bundle creator able to define its privacy data processings clear and defined in a standard way.

if you experiment this as a separate bundle and component first, you still have all the same tools available to allow other bundles to provide some infos (FrameworkBundle is a bundle like any others).
And experimenting separately gives you much more flexibility to try different things: once a component is merged into Symfony, our BC promise makes it hard to perform big changes if you find out that the component does not work as is. And you will only start getting feedback on it after it is part of a stable release that people can use (so months after working on it). For an external package, you can start making releases much faster. So creating this directly as a Symfony would probably take a long time because of the need to get it right (or almost right) in the first iteration.
That's the main reason I push for starting this externally first (and discussing after that whether including it in core is worth it or no): it allows experimenting to reach the right solution.

To give you a comparison, the same happens for the work on PSR components. PSR-11 was released 1 year ago. but the prototyping in the container-interop github organization more than 4 years ago. Trying to get something right without any prototyping is quite hard.

**Comment by Aerendir** [comment]
@stof , yes, I completely agree on what you think about: for this reason I created the repository and I will start working in it.

For the moment I'm reorganizing ideas to spot the possible things possible in the privacy field.

I've given some examples above, but may be there other use cases that can be covered... We'll see :)

**Comment by CoalaJoe** [comment]
@Aerendir Do you have a link to the repo? Or need help with something?

**Comment by fabpot** [comment]
I'm going to close this issue as I agree with @stof, it's probably something that can be developed outside of Symfony first.
